ql/src/java/org/apache/hadoop/hive/ql/lockmgr/DbTxnManager.java

```diff
@@ -485,6 +493,27 @@ private void clearLocksAndHB() {
     stopHeartbeat();
   }
 
+  private void cleanupOutputDir(Context ctx) throws MetaException {
+    if (HiveConf.getBoolVar(conf, HiveConf.ConfVars.TXN_CTAS_X_LOCK)) {
+      Table destinationTable = ctx.getDestinationTable();
+      if (destinationTable != null) {
+        try {
+          CompactionRequest rqst = new CompactionRequest(
+                  destinationTable.getDbName(), destinationTable.getTableName(), CompactionType.MAJOR);
+          rqst.setRunas(TxnUtils.findUserToRunAs(destinationTable.getSd().getLocation(),
+                  destinationTable.getTTable(), conf));
+
+          rqst.putToProperties(META_TABLE_LOCATION, destinationTable.getSd().getLocation());
+          rqst.putToProperties(IF_PURGE, Boolean.toString(true));
+          TxnStore txnHandler = TxnUtils.getTxnStore(conf);
```

Comment:
that would be expensive to create a new txnStore everytime

Reply:
btw, would it be hard to create a completionHook similar to Iceberg one?

Reply:
> btw, would it be hard to create a completionHook similar to Iceberg one?

We could create one but it would include failures only within Query execution.
Anything done after query execution (post execution activities like releasing locks) will not be within its scope, which is why I disregarded the Hook approach.

The hooks are used as part of finally block here - 
https://github.com/apache/hive/blob/b197ed86029f07696e326acb5878f86c286e9e1a/ql/src/java/org/apache/hadoop/hive/ql/Executor.java#L118

Cleanup will then be dependent on a HiveConf - `hive.query.lifetime.hooks`.